### ISSUE TITLE
Scanner improvements

### ISSUE THREAD

**Issue by ilammy** [ilammy/sash#1]
I have an idea of writing a Sash plugin for _IntelliJ IDEA_ to get IDE support almost for free. The first step to this would be a scanner, so that we have at least basic syntactic highlighting and stuff. We have a scanner, but it certainly needs improvements before we can start using its code.
- [x] ~~**Documentation.** Lexical syntax of Sash should be documented, as well as have better technical documentation~~
  - [x] ~~_Lexical syntax:_ documentation in `doc` directory at root level.~~
  - [x] _Implementation:_ outline should be described in `src/README.md`
- [x] **Library.** The crate should be turned into a library, like _libsyntax_ or something.
- [x] **Refactoring.** I am not very satisfied with some points in the code and they definitely need some improvement.
  - [x] _Readability:_
    - [x] Better assertions for expected character sequences
    - [x] Additional comments
    - [x] Rustic style
  - [x] ~~_Tracing:_ add a bunch of `debug!` calls to trace what happens during scanning.~~
  - [x] _Unit tests:_
    - [x] Avoid boilerplate ~~and arbitrary calculations~~
    - [x] Better failure messages
    - [x] ~~Check for the kind of errors detected~~
  - [x] _Structure:_ split out token type definitions and diagnostics into separate modules.
- [x] **Improvements.** Some additional features of the scanner that need to be looked into.
  - [x] _Diagnostics:_
    - [x] Review and unify error messages.
    - [x] Improve diagnostic levels. They should be clear to use. Look at Clang or whatever.
    - [x] Better interface for diagnostic reporting.
  - [x] _Context:_ the scanner should provide as much information about tokens as possible. E.g., it should tell the type suffix of a literal.
  - [x] _Interpreting tokens:_ the lexer should do some interpreting of the tokens. E.g., ~~normalize identifiers,~~ expand escape sequences in strings. ~~Maybe parse numbers into their values.~~
  - [x] _Coverage:_ I am absolutely sure that I have missed some cases in unit-tests. They need to be improved for better coverage. [Here is how it can be done with kcov.](https://users.rust-lang.org/t/tutorial-how-to-collect-test-coverages-for-rust-project/650) (I do not think that Rust tooling is mature enough to include coverage reports into CI, but it is okay for occasional manual runs.)
- [x] **TODOs.** There are quite a few sprinkled all over the code. They need to be resolved.
- [x] **Identifier normalization.** Normalize identifiers and implicit symbols as a part of scanning.
  - [x] Implement Unicode normalization algoritms.
  - [x] Integrate them into scanner.
- [x] ~~**Reserved identifiers.** Integrate reserved identifeirs into tokens produced by the scanner.~~

**Comment by ilammy** [comment]
Well... I spent a couple of days trying to integrate this scanner with IDEA for syntax highlighting via JNI. It turned out to be not so straightforward as I thought. It may even be better to write the whole lexer in Java rather than trying to reuse this one, at least for the IDE. I am not so sure now in the future steps for the IDE support.

So much for the IDE, there's still a parser for compiler to write. I can deal with the IDE later. Probably. Though, I really don't know about any relatively successful usages of compiler's API in IDEs aside from libclang usage in Xcode. However, I still think that this is the _right way_ to do it; albeit maybe not the fastest one.

Anyway... I guess I will just focus on Sash compiler itself for now.

**Comment by ilammy** [comment]
After giving it some thought I decided to postpone the interpretation as well. First of all, this is a scanner, it should only provide _locations_ of tokens, not bother with their actual semantics. Then, take floats for example. Without an explicit type suffix we are not sure whether we need single or double precision for a literal; and we don't want to lose precision because of that. Finally, we will need precise as-in-the-source forms for reporting errors anyway. We can always quickly parse correct tokens somewhere in the parser before emitting them into AST.

**Comment by ilammy** [comment]
Getting back to this...

After some thought I believe that scanner _should_ normalize identifiers and implicit symbols. This is similar to how it expands escape sequences, so there is not much sense into separating this concern.

And then, I believe Sash will have reserved keywords so they also should be separated from identifers during the scanning phase.

**Comment by ilammy** [comment]
Documentation will be done together with the formal grammar. There is no real need for it before the syntax is stabilized. The same for the reserved keywords: I have no idea which ones we'll need as there is no parser yet. We will think about along the way. It would be pretty easy to integrate them into identifier scanning code. So, I'm finally closing this issue.
